refactor(spell): migrate to Lua, drop netrw dependency

tampueroc opened the pull request:
## Problem

Disabling netrw breaks the spellfile auto-download feature.

## Solution

Migrate the spell file download logic from Vimscript to Lua, removing the dependency on netrw. 
- Add tests
- Use `vim.net.request()` for downloading spellfiles instead of `:Nread` (netrw).

Closes #7189

tampueroc committed:
refactor(spell): migrate to Lua, drop netrw dependency

Problem:
Spell file downloads relied on Vimscript and netrw (:Nread). If netrw is
disabled, downloads fail.

Solution:
Port the logic to Lua as `nvim.spellfile` and wire it via a Lua plugin that
handles `SpellFileMissing`. Use `vim.net.request()` with a timeout for HTTP,
prompt via `vim.fn.input` and report via `vim.notify`.

Closes #7189

tampueroc commented:
> This is way too much code. Please choose the most high-value tests and drop the rest. And we 100% aren't going to add 4 new test files just for spell download.

Reasonable...trimmed to one spec (`spellfile_spec.lua`) with 3 high-value offline tests (no operation, download success, dual fail) and dropped the rest.

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,166 @@
+
+describe('vim.spellfile', function()
+  before_each(function()
+    n:clear()
```
```suggestion
    n.clear()
```

justinmk commented on runtime/lua/vim/spellfile.lua:
This file should live in `runtime/lua/nvim/`. It's not part of the `vim.*` stdlib, it's just a plugin.

justinmk commented on runtime/lua/vim/spellfile.lua:
```diff
@@ -0,0 +1,281 @@
+M.config = {
+  url = 'https://ftp.nluug.nl/pub/vim/runtime/spell',
+  rtp = vim.opt.rtp:get(),
+  encoding = 'utf-8',
```
The old code referenced `&encoding` because the 'encoding' option of Vim is configurable. In Nvim it's always utf-8. So I think we don't need this parameter at all.

If you want to keep the logic as-is, you can set `local encoding = 'utf-8'` at the top of this module. But it shouldn't be a public config setting.

justinmk commented on runtime/lua/vim/spellfile.lua:
```diff
@@ -0,0 +1,281 @@
+}
+
+---@type table<string, boolean>
+M.done = {}
```
why is this on the public interface?

tampueroc replied:
Solved, would a getter like `getDone` be good? I'm accessing this attr in `spellfile_spec.lua`.

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,166 @@
+      return { warns = warns, done = done, reloaded = reloaded }
+    ]])
+
+    assert.are.same(1, out.warns)
+    assert.is_true(out.done)
+    assert.is_false(out.reloaded)
```
use `t.eq` and `t.ok`, don't use the `assert` module directly

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,166 @@
+      vim.fn.isdirectory = function(_) return 0 end
+      vim.fn.stdpath = function(k) assert(k=='data'); return '/tmp' end
+      vim.fn.mkdir = function(_) return 1 end
+      if vim.uv then
```
where did this come from? `vim.uv` is always available.

if this code is generated by AI, please review it yourself before asking us to review.

tampueroc replied:
Removed, thank you! it was redundant...

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,166 @@
+
+    assert.is_true(out.spl_present)
+    assert.is_false(out.sug_present)
+    assert.is_true(out.reloaded)
```
this probably came from AI. look at existing test files to understand the codebase conventions.

tampueroc replied:
Is the issue the assert pattern or the method? I updated `spellfile_spec.lua` to use `t.eq`.

tampueroc commented:
Pushed an update: proper rtp setup in tests (I tried to mirror `runtime_spec.lua`), encoding fixed to `utf-8`, switched assertions to `t.eq`, removed redundant code in plugin and tests (`vim.loop`....). Should look better but let me know if you prefer a different pattern for the testing suite! 

FYI I mirrored [#29552](https://github.com/neovim/neovim/pull/29552) a bit too closely before, tests are now rewritten to match our conventions.

tampueroc commented:
@justinmk For the sake of transparency: I did use AI assistance to help with the test cases, mainly reviewing them afterward to understand intent and make sure they ran correctly with what I expected. I should have signaled that earlier so you knew what to expect, sorry for wasting honest review efforts...I believe I incorporated the feedback as best as I can now. Just wanted to be upfront and thank you for your help!

justinmk commented on runtime/lua/nvim/spellfile.lua:
```diff
@@ -0,0 +1,278 @@
+  return M._done[key]
+end
+
+function M.setup(opts)
```
does not look used?

justinmk commented on runtime/lua/nvim/spellfile.lua:
```diff
@@ -0,0 +1,278 @@
+  return vim.opt.rtp:get()
+end
+
+function M.isDone(key)
```
i dropped this.

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,160 @@
+    local out = exec_lua([[
+      local s = require('nvim.spellfile')
+
+      local data_root = 'Xdata2'
```
is this "2" suffixed because the other test was not cleaning up `Xdata/` ?

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,160 @@
+
+      vim.net.request = orig_req
+      vim.notify = orig_notify
+      vim.cmd = orig_cmd
```
don't need to restore these because each test starts a new Nvim process (`n.clear()`)

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,160 @@
+
+    local requests = 0
+    local orig_req = vim.net.request
+    vim.net.request = function(...) requests = requests + 1 end
```
If the plugin switches to something other than `vim.net.request`, this will will just continue passing. Example of why mocks are really discouraged.

justinmk commented on runtime/lua/nvim/spellfile.lua:
```diff
@@ -0,0 +1,278 @@
+  local sug_name = string.format('%s.%s.sug', lang, enc)
+
+  local url_utf8 = M.config.url .. '/' .. spl_utf8
+  local out_utf8 = vim.fs.joinpath(dir, spl_utf8)
```
when I started with `NVIM_APPNAME=foo nvim` and tried `set spelllang=fr` i see: 

```
No spell file found for fr (utf-8). Download? [y/N] yDownloading fr.utf-8.spl …
Saved fr.utf-8.spl to <runtimepath>/neovim/runtime/spell/fr.utf-8.spl
Downloading fr.utf-8.sug …
Saved fr.utf-8.sug to <runtimepath>/neovim/runtime/spell/fr.utf-8.sug
```

why didn't it create a `spell/` dir in the user-local stdpath()?

justinmk commented on runtime/lua/nvim/spellfile.lua:
```diff
@@ -0,0 +1,278 @@
+  if #dirs > 0 then
+    return dirs[1]
+  end
+  local target = default_spell_dir()
```
this should prefer a `stdpath()` dir. looking for "writable dirs in runtimepath" should be a last resort.

I assume this was matching the old logic, but makes more sense to prefer stdpath()

justinmk commented on runtime/lua/nvim/spellfile.lua:
```diff
@@ -0,0 +1,278 @@
+  return opts
+end
+
+function M.choose_directory()
```
this is not actually used anywhere

justinmk committed:
refactor(spell): cleanup

- prefer `stdpath(data)/site/spell` instead of looking for random dirs in 'runtimepath'.
- drop unused functions `choose_directory`, `setup`, etc.

justinmk commented on runtime/lua/nvim/spellfile.lua:
```diff
@@ -0,0 +1,278 @@
+      vim.schedule(function()
+        vim.cmd('echo ""')
+      end)
+      M._done[info.key] = true
```
Seems like an alternative to `M._done` would be to accept callbacks in `download`:

    download(..., on_done, on_err)

justinmk commented on test/functional/lua/spellfile_spec.lua:
```diff
@@ -0,0 +1,160 @@
+      local old_access = vim.uv.fs_access
+      vim.uv.fs_access = function(_, _) return false end
+      local old_stat = vim.uv.fs_stat
+      vim.uv.fs_stat = function(p) return old_stat and old_stat(p) or nil end
```
neither of these seems relevant to the test.

justinmk reviewed (approved):
Thanks for getting this started! This is helpful and part of our roadmap. 

I pushed a few updates. Next time please try to look carefully at the generated code, and avoid mocks.

When testing, I noticed a bug, but the bug is present on master also: after downloading the spell file, it doesn't "take effect" even after `:set spell`, until after restarting Nvim. And it shows this:

    Warning: Cannot find word list "fr.utf-8.spl" or "fr.ascii.spl"
